qiskit_nature/operators/second_quantization/fermionic_op.py

```diff
@@ -249,157 +231,112 @@ def mul(self, other: complex) -> "FermionicOp":
-
-        # count the number of `+` and `-` in the first label ahead of time
-        count = label1.count("+") + label1.count("-")
-
-        for pair in zip(label1, label2):
-            # update the count as we progress
-            char1, char2 = pair
-            if char1 in "+-":
-                count -= 1
-
-            new_char = cls._MAPPING[pair]
-            if new_char == "0":
-                # if the new symbol is a zero-op, return early
-                return "I" * len(label1), 0
-            new_label.append(new_char)
-            # NOTE: we can ignore the type because the only scenario where an `int` occurs is caught
-            # by the `if`-statement above.
-
-            # If char2 is one of `+` or `-` we pick up a phase when commuting it to the position
-            # of char1. However, we only care about this if the number of permutations has odd
-            # parity.
-            if count % 2 and char2 in "+-":
-                sign *= -1
-
-        return "".join(new_label), sign
+        new_data = list(
+            filter(
+                lambda x: x[1] != 0,
+                (
+                    (" ".join(filter(None, (label1, label2))), cf1 * cf2)
+                    for label2, cf2 in other._data
+                    for label1, cf1 in self._data
+                ),
+            )
+        )
+        register_length = max(self.register_length, other.register_length)
+        if not new_data:
+            return FermionicOp(("", 0), register_length)
+        return FermionicOp._from_legacy(FermionicOp(new_data, register_length)._to_legacy())
 
     def add(self, other: "FermionicOp") -> "FermionicOp":
         if not isinstance(other, FermionicOp):
             raise TypeError(
                 f"Unsupported operand type(s) for +: 'FermionicOp' and '{type(other).__name__}'"
             )
 
-        # Check compatibility (i.e. operators act on same register length)
-        if self.register_length != other.register_length:
-            raise TypeError("Incompatible register lengths for '+'.")
-
         return FermionicOp(
-            list(
-                zip(
-                    self._labels + other._labels,
-                    np.hstack((self._coeffs, other._coeffs)).tolist(),
-                )
-            )
+            self._data + other._data,
+            max(self.register_length, other.register_length),
         )
 
-    def to_list(self) -> List[Tuple[str, complex]]:
+    # pylint: disable=arguments-differ
+    def to_list(self, sparse: bool = False) -> List[Tuple[str, complex]]:
         """Returns the operators internal contents in list-format.
 
+        Args:
+            sparse: The labels in the return list are sparse or not. (Default: False)
+
         Returns:
             A list of tuples consisting of the dense label and corresponding coefficient.
         """
-        return list(zip(self._labels, self._coeffs.tolist()))
+        if sparse:
+            return self._data[:]
```

Comment:
```suggestion
            return self._data
```

Reply:
This `[:]` means copy of list in Python. I think this copy is necessary because of immutability.

Reply:
Ah I did not know that. A comment would be useful there in order to avoid future confusion.
Alternatively, why not call `.copy()` directly?

Reply:
Both are often used, and there is no deep meaning to my choice. `.copy()` is equivalent to `[:]`. https://docs.python.org/3/tutorial/datastructures.html#more-on-lists
 I will change `.copy()` here.